src/Reopt/TypeInference/ConstraintGen.hs

```diff
@@ -129,165 +131,226 @@ newtype Warning = Warning String
 instance Show Warning where
   show (Warning w) = w
 
-type PhiType arch = Map (Some (FnPhiVar arch)) IType
+-- | Context available throughout all of the constraint generation
+newtype CGenGlobalContext arch = CGenGlobalContext
+  { _cgenMemory      :: Memory (ArchAddrWidth arch)
+  }
 
-data CGenContext arch = CGenContext {
-  -- Global state
-  cgenFunTypes        :: Map (ArchSegmentOff arch) (FunType arch)
-  -- | External functions (only named).
-  , cgenExtFunTypes   :: Map BSC.ByteString (FunType arch)
+makeLenses ''CGenGlobalContext
```

Comment:
Do we ever update these types?  If they are read only I don't see that we need lenses

Reply:
True!